sdks/python/apache_beam/io/parquetio.py

```diff
@@ -0,0 +1,349 @@
+    source_from_file = partial(
+        _create_parquet_source,
+        min_bundle_size=min_bundle_size,
+        columns=columns
+    )
+    self._read_all_files = filebasedsource.ReadAllFiles(
+        True, CompressionTypes.AUTO, desired_bundle_size, min_bundle_size,
+        source_from_file)
+
+    self.label = label
+
+  def expand(self, pvalue):
+    return pvalue | self.label >> self._read_all_files
+
+
+def _create_parquet_source(file_pattern=None,
+                           min_bundle_size=None,
+                           validate=False,
+                           columns=None):
+  return \
+    _ParquetSource(
+        file_pattern=file_pattern,
+        min_bundle_size=min_bundle_size,
+        validate=validate,
+        columns=columns
+    )
+
+
+class _ParquetUtils(object):
+  @staticmethod
+  def find_first_row_group_index(pf, start_offset):
+    for i in range(_ParquetUtils.get_number_of_row_groups(pf)):
+      row_group_start_offset = _ParquetUtils.get_offset(pf, i)
+      if row_group_start_offset >= start_offset:
+        return i
+    return -1
+
+  @staticmethod
+  def get_offset(pf, row_group_index):
+    first_column_metadata =\
+      pf.metadata.row_group(row_group_index).column(0)
+    if first_column_metadata.has_dictionary_page:
+      return first_column_metadata.dictionary_page_offset
+    else:
+      return first_column_metadata.data_page_offset
+
+  @staticmethod
+  def get_number_of_row_groups(pf):
+    return pf.metadata.num_row_groups
+
+
+class _ParquetSource(filebasedsource.FileBasedSource):
+  """A source for reading Parquet files.
+  """
+  def __init__(self, file_pattern, min_bundle_size, validate, columns):
+    super(_ParquetSource, self).__init__(
+        file_pattern=file_pattern,
+        min_bundle_size=min_bundle_size,
+        validate=validate
+    )
+    self._columns = columns
+
+  def read_records(self, file_name, range_tracker):
+    next_block_start = -1
+
+    def split_points_unclaimed(stop_position):
+      if next_block_start >= stop_position:
+        # Next block starts at or after the suggested stop position. Hence
+        # there will not be split points to be claimed for the range ending at
+        # suggested stop position.
+        return 0
+      return RangeTracker.SPLIT_POINTS_UNKNOWN
+
+    range_tracker.set_split_points_unclaimed_callback(split_points_unclaimed)
+
+    start_offset = range_tracker.start_position()
+    if start_offset is None:
+      start_offset = 0
+
+    with self.open_file(file_name) as f:
+      pf = ParquetFile(f)
+
+      index = _ParquetUtils.find_first_row_group_index(pf, start_offset)
+      if index != -1:
+        next_block_start = _ParquetUtils.get_offset(pf, index)
+      else:
+        next_block_start = range_tracker.stop_position()
+      number_of_row_groups = _ParquetUtils.get_number_of_row_groups(pf)
+
+      while range_tracker.try_claim(next_block_start):
+        table = pf.read_row_group(index, self._columns)
```

Comment:
How large can a row group be ? Can we run into OOMs at this point ?

Reply:
Typical size of a row group for hdfs stored parquet is a few hundred megabytes (128~256MB) I assume. We can run into OOM if the size of a row group is very big but that might not be very common case since writing a large row group requires the same big amount of memory as you read it.